draft-ietf-jsonpath-base.md

```diff
@@ -1346,20 +1349,24 @@ Note that `..` on its own is not a valid segment.
 
 A descendant segment produces zero or more descendants of the input value.
 
-A descendant selector of the form `..[<selectors>]` visits each node of the input value and
-its descendants in such an order that:
+A descendant selector visits the input value and each of
+its descendants such that:
 
 * nodes of any array are visited in array order, and
-* nodes are visited before all their descendants.
-
-It applies the child segment `[<selectors>]` to each node and concatenates the resultant nodelists together
-in the order in which the nodes were visited.
+* nodes are visited before their descendants.
 
-This definition does not stipulate the order in which the children of an object are visited, since
+The order in which the children of an object are visited is not stipulated, since
 JSON objects are unordered.
 
-Where a selector can produce a nodelist in more than one possible order, the selector may produce nodelists in distinct
-orders each time it appears in the descendant segment.
+Suppose the nodes, in the order visited, are `D1`, ..., `Dn` (where `n >= 1`).
+Note that `D1` is the input value.
+
+For each `i` such that `1 <= i <= n`, the nodelist `Ri` is defined to be a result of applying
+the child segment `[<selectors>]` (with all its selectors) to the node `Di`.
+Note that `Ri` is the concatenation of the nodelists from each of the selectors.
+
+The result of the descendant selector is the concatenation of `R1`, ..., `Rn` (in that order).
+Note that only a single traversal of the input value and its descendants is required.
```

Comment:
```suggestion
Suppose the nodes, in the order visited, are `D1`, ..., `Dn` where `n >= 1`.
(Note that `D1` is the input value.)

For each node `Di` such that `1 <= i <= n`, the resulting nodelist `Ri` is defined to be a result of applying
the all of the selectors in the child segment `[<selectors>]` to the node `Di`.

The final result of the descendant selector is the concatenation of `R1`, ..., `Rn` (in that order).
```

I fear that

> Note that only a single traversal of the input value and its descendants is required.

may be getting too close to specifying implementation logic.

Reply:
Ok, I'll delete the single traversal sentence.

I didn't adopt the proposal because it talks about applying all the selectors in the child segment, but we need child segment semantics (specifically, concatenating the selector nodelists) instead of just selector semantics.

Reply:
My suggestion also provides a more precise wording and adjusts the parenthetical phrases.

> it talks about applying all the selectors in the child segment, but we need child segment semantics (specifically, concatenating the selector nodelists) instead of just selector semantics

Then I suggest we remove mention of the selectors completely.

```
Suppose the nodes, in the order visited, are `D1`, ..., `Dn` where `n >= 1`.
(Note that `D1` is the input value.)

For each node `Di` such that `1 <= i <= n`, the intermediate nodelist `Ri` is the result of applying
the child segment `[<selectors>]` to the node `Di`.

The result of the descendant segment is the concatenation of the intermediate nodelists `R1`, ..., `Rn` (in that order).
```

We also need to be sure that in the "child segment" section, we define how the selector results are concatenated, which it seems we do:

> The resulting nodelist of a child segment is the concatenation of the nodelists from each of its selectors in the order that the selectors appear in the list.

Reply:
> Then I suggest we remove mention of the selectors completely.

I can live with that. See 3516bee163b81457e64be5dabe2d417893d38bbb.

Reply:
Why are you not taking the rest of the suggestion?

Reply:
Ah, sorry, I should also have deleted "Note that `Ri` is the concatenation of the nodelists from each of the selectors.". Fixed in 0d25c700fb7e02ec7db97bf3acc3016085384fe5.

As for the rest of the suggestion, I prefer the current wording in the PR:
* I prefer to define `Ri` via universal quantification over `i`, rather than over `Di`.
* I don't see any advantage in introducing the extra term "intermediate nodelist".

Are you comfortable with approving now?

Reply:
Thanks for explaining your opinions. I beg to differ. Is this a blocker for you or can you now approve the PR?